File: 1-two-sum/python/optimal.py
from typing import List

class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        hash_map = {}
        for i in range(len(nums)):
            number_to_find = target - nums[i]

            if number_to_find in hash_map:
                # return [i, hash_map[number_to_find]]
                return [nums[i], nums[hash_map[number_to_find]]]
            hash_map[nums[i]] = i

s = Solution()

print(s.twoSum([2, 7, 11, 15], 9))

File: 1-two-sum/js/optimal.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
const twoSum = function (nums, target) {
  const hashMap = {};
  for (let i = 0; i < nums.length; i++) {
    const numberToFind = target - nums[i];
    if (hashMap.hasOwnProperty(numberToFind)) {
    //   return [i, hashMap[numberToFind]];
      return [nums[i], nums[hashMap[numberToFind]]]; // if we need to return number instead of value
    }
    hashMap[nums[i]] = i;
  }
};

console.log(twoSum([2, 7, 11, 15], 9));
